### PR TITLE
feat: add model script, training recipe and pretrained weight of halonet

### PR DESCRIPTION
**Pull request by xjtu-fgh**
Thank you for your contribution to the MindCV repo.
Before submitting this PR, please make sure:

- [ ] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [ ] Your code builds clean without any errors or warnings
- [ ] You are using approved terminology
- [ ] You have added unit tests

## Motivation

(Write your motivation for proposed changes here.)

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: configs/halonet/README.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# HaloNet
+> [Scaling Local Self-Attention for Parameter Effificient Visual Backbones](https://arxiv.org/abs/2103.12731)
+
+## Introduction
+
+​		Researchers from Google Research and UC Berkeley have developed a new model of self-attention that can outperform standard baseline models and even high-performance convolutional models.[[1](#references)]
+​		Blocked Self-Attention：The whole input image is divided into multiple blocks and self-attention is applied to each block.However, if only the information inside the block is considered each time, it will inevitably lead to the loss of information.Therefore, before calculating the SA, a haloing operation is performed on each block, i.e., outside of each block, the information of the original image is used to padding a circle, so that the sensory field of each block can be appropriately larger and focus on more information.
```
Comment (Songyuanwei) [inline configs/halonet/README.md:7]: 换行好像没起作用

File: configs/halonet/README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+## Results
+
+Our reproduced model performance on ImageNet-1K is reported as follows.
+
+<div align="center">
+
+| Model      | Context  | Top-1 (%) | Top-5 (%) | Params (M) | Recipe                                                       | Download                                                     |
+| ---------- | -------- | --------- | --------- | ---------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| halonet50t | D910X8-G | 79.53     | 94.79     | 28.59      | [yaml](https://github.com/mindspore-lab/mindcv/blob/main/configs/halonet/halonet_50t_ascend.yaml) | [weights](https://download.mindspore.cn/toolkits/mindcv/halonet/halonet_50t-533da6be.ckpt) |
```
Comment (Songyuanwei) [inline configs/halonet/README.md:19]: 名字改成 halonet_50t

Comment (Songyuanwei) [inline configs/halonet/README.md:19]: 模型参数是28.59吗？应该不是28.59. 看一下训练日志里面打印的模型参数信息

File: configs/halonet/halonet_50t_ascend.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+opt: 'adamw'
+filter_bias_and_bn: True
+weight_decay: 0.02
+loss_scale: 1024
+use_nesterov: False
+
+# lr scheduler
+scheduler: 'warmup_cosine_decay'
+min_lr: 0.000006
+lr:  0.000234375
```
Comment (Songyuanwei) [inline configs/halonet/halonet_50t_ascend.yaml:53]: lr好像不是这个吧。。0.00125？确认一下，其他参数也对照确认一下

File: mindcv/models/halonet.py [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+        "url": url,
+        "num_classes": 1000,
+        "first_conv": "stem.conv1.conv",
+        "classifier": "classifier",
+        **kwargs,
+    }
+
+
+default_cfgs = {
+    "halonet50t": _cfg(
```
Comment (Songyuanwei) [inline mindcv/models/halonet.py:30]: 应该是halonet_50t吧。参考edgenext.py，_cfg里面加一下input_size=(3, 256, 256)

File: mindcv/models/halonet.py [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+        out_stage4 = self.stage4(out_stage3)
+        out = self.classifier(out_stage4)
+        return out
+
+
+@register_model
+def halonet_50t(pretrained: bool = False, num_classes: int = 1000, in_channels=3, **kwargs):
+    """Get HaloNet model.
+    Refer to the base class `models.HaloNet` for more details."""
+    default_cfg = default_cfgs["halonet"]
```
Comment (Songyuanwei) [inline mindcv/models/halonet.py:628]: 和上面default_cfgs里面的名字对齐

File: mindcv/models/halonet.py [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+        use_avg_pool = False
+        if stride > 1:
+            use_avg_pool = avg_down or block_size % stride != 0
+            self.block_stride = stride
+            self.block_size_ds = self.block_size // self.block_stride
+        self.q = nn.Conv2d(dim,
+                           self.dim_out_qk,
+                           1,
+                           stride=self.block_stride,
+
```
Comment (Songyuanwei) [inline mindcv/models/halonet.py:283]: 这行空行是什么。。没用的话去掉

File: mindcv/models/halonet.py [hunk 7]
```diff
@@ -0,0 +1,647 @@
...
+        self.conv1 = ConvBnAct(3, 24, kernel_size=3, stride=2, padding=1, act=act)
+        self.conv2 = ConvBnAct(24, 32, kernel_size=3, stride=1, padding=1, act=act)
+        self.conv3 = ConvBnAct(32, 64, kernel_size=3, stride=1, padding=1, act=act)
+        self.pool = nn.MaxPool2d(kernel_size=3, stride=2, pad_mode='valid')
+
+    def construct(self, x):
+        x = self.conv1(x)
+        x = self.conv2(x)
+        x = self.conv3(x)
+        x = ms.numpy.pad(x, ((0, 0), (0, 0), (1, 0), (1, 0)), constant_values=-32768)
```
Comment (Songyuanwei) [inline mindcv/models/halonet.py:150]: 这里就用你之前提交的代码。。init里面初始化self.pad， construct里面调用。。不要有这些奇奇怪怪的数字

File: configs/halonet/README.md
Comment (Songyuanwei) [inline configs/halonet/README.md:1]: 还有之前关掉的PR里面很多修改没有完成

File: configs/halonet/README.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
+HaloNet
```
Comment (Songyuanwei) [inline configs/halonet/README.md:1]: 这个是标题，要大写的。看一下其他人的写法

File: mindcv/models/halonet.py [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+    :param permute_mask: permute output axis according to this
+    """
+    B, H, W, _ = q.shape
+    rel_size = rel_k.shape[0]
+    win_size = (rel_size+1)//2
+    rel_k = ops.transpose(rel_k, (1, 0))
+    x = msnp.tensordot(q, rel_k, axes=1)
+    x = ops.reshape(x, (-1, W, rel_size))
+    # pad to shift from relative to absolute indexing
+    pad = ops.Pad(paddings=((0, 0), (0, 0), (0, 1)))
```
Comment (geniuspatrick) [inline mindcv/models/halonet.py:153]: 这个可以用函数式接口吗？

**Review by Songyuanwei — APPROVED** [review]

**Review by geniuspatrick — APPROVED** [review]

**Pull request merged**